StandardConnectionGadget: shift click connection

mattigruener opened the pull request:
Hey John,

we had talked about adding the shift-click feature on that other [PR](https://github.com/GafferHQ/gaffer/pull/2072) and here's a first prototype that will still need some fixes. I'm sure you'll find more issues, but these are the two bigger things that I'd like to have a quick discussion about:

1. Bringing over the code that cleans up the old connection to the `StandardConnectionGadget` was generally possible, but we need a way to determine if we have actually connected back to the plug that was connected already. That was easy on the Nodule side of things, but as far as I can see there's no way for the ConnectionGadget to know what Nodule we dragged onto? Maybe I'm missing something that we store in the event. What I've done now is to compare the list of outputs of the source plug before and after dragging. If the two are the same, we shouldn't clean up anything. Not sure how you feel about storing that OutputContainer - intuitively I'd think that there should be a cleaner solution. I reset that Container to an empty list once we're done so that we don't hold on to anything we don't need anymore, but it still feels messy to me.

2. I looked into getting undo to behave properly. Spreading functionality into two classes requires merging the reconnecting and the cleaning up on the undo stack. I've used the mergeGroup argument for that, but if we successively change multiple connections, they all get merged into just one item on the stack now. I guess what I would need is an UndoScope that does something like "use this mergeGroup, but don't merge into existing elements on the stack. Only be ready for subsequent actions to be merged on top of you" - if that makes sense? I'd put that in the StandardNodule, which would start a new element on the stack every time a `drop()` occurs on the Nodule. Maybe I'm missing something and we have something like that already? If we don't - I'm happy to look into adding it, but I wanted to discuss with you first.

In case you don't like the idea of moving that functionality over anymore after having seen what it ends up as, just let me know. I can try to come up with something else. I guess moving it over makes things a little more complicated, but is generally the right thing to do in terms of decoupling the two classes. I guess we can't have the Nodule just add itself to the event and have the StandardConnectionGadget handle setting the input, as well? I've actually not looked into that too closely, but I guess the `data` member of the event is `const` so we can't just sneak in something else to communicate back to the StandardConnectionGadget?

Thanks :)

Matti.

johnhaddon commented:
Hi Matti,

Thanks for looking into this - it's all a bit thorny isn't it? In general it seems that the GraphGadget's constituent parts need to collaborate more than we might like to implement some of the interaction we want. It's bad design on my part really, but let's see if we can muddle through with what we have for now...

> as far as I can see there's no way for the ConnectionGadget to know what Nodule we dragged onto?

There is `DragDropEvent::destinationGadget` actually - you can see this being used in `StandardConnectionGadget::dragEnd()`. Perhaps this is the key to coming up with something workable.

> What I've done now is to compare the list of outputs of the source plug before and after dragging. If the two are the same, we shouldn't clean up anything. I reset that Container to an empty list once we're done so that we don't hold on to anything we don't need anymore, but it still feels messy to me.

Yeah, the more state things store, the harder it is to keep everything in sync and robust. Feels messy to me too...

> Spreading functionality into two classes requires merging the reconnecting and the cleaning up on the undo stack. I've used the mergeGroup argument for that, but if we successively change multiple connections, they all get merged into just one item on the stack now. I guess what I would need is an UndoScope that does something like "use this mergeGroup, but don't merge into existing elements on the stack. Only be ready for subsequent actions to be merged on top of you" - if that makes sense?

Yeah, that does make sense, as a logical consequence of splitting the responsibility so that Nodule does half the work and ConnectionGadget does the rest. But now I've seen it in practice, I think the splitting was a very poor suggestion on my part. I don't think we should add API features to encourage this sort of awkward coupling.

> In case you don't like the idea of moving that functionality over anymore after having seen what it ends up as, just let me know. I can try to come up with something else. I guess moving it over makes things a little more complicated, but is generally the right thing to do in terms of decoupling the two classes.

I would like to decouple things as much as possible, but I'm beginning to think that it's not totally possible in the case of connection/nodule interaction, and what's actually missing is another abstraction to provide a cleaner coupling. When I first wrote the GraphGadget code, it seemed totally logical that connections/nodules would start drags with the plug as the source data, and a nodule would receive plugs in drop and set up the connection. And in fact that worked pretty well for a while, but these features came along to confound it :

- PlugAdders. These need to start a drag, but don't yet have a plug. There's a special case in `StandardNodule::drop()` to deal with that, and it's actually the PlugAdder that does the heavy lifting when the drag ends, not the Nodule.
- Endpoint snapping. When a connection is dragged onto a Nodule, we want to snap the endpoint and tangent of the connection to it, to provide a visual clue to the user. This resulted in the `updateDragEndPoint()` mess in `StandardNodule::dragEnter()`, where there's a different special case to call the appropriate function on three different types of drag source.
- Shift+dragging of connections. The one we're dealing with now.

What all these things have in common is that they're not just dropping a plug onto a Nodule, they're manipulating a connection, be it an already-existing one or a virtual about-to-be-made one. So I think what is actually missing is an abstraction for "connection creators" that the Nodule can interact with, so there's just a single special case for interacting with those, and then multiple concrete implementations of them behind that.

I'm not quite sure what that should look like yet though, and when I'm not sure I like to muddle through until we have enough use cases for it to become clear.

> I guess we can't have the Nodule just add itself to the event and have the StandardConnectionGadget handle setting the input, as well?

So we'd do it all in `StandardConnectionGadget::dragEnd()`? using `event.destinationGadget`? That sounds like a great suggestion to me, and it also somewhat matches what happens with the PlugAdder right now, which bodes well for refactoring this all into some sort of "connection creator" abstraction at some later point in the future. Sorry to have led you on a wild goose chase with my original suggestion, and thanks for sticking with it...

Cheers...
John

mattigruener commented:
Cool, I'll give that a try today. Thanks!

mattigruener commented:
Urg, without having tried yet - I guess what I was missing when making the suggestion to handle it all on the connection side was that we will have to have a connection for that to work. It'll be fine if we just move the connection over to a different plug, but making the first connection always needs to be handled in the Nodule, right?

mattigruener commented:
One more thing that I've noticed: The event.destinationGadget doesn't reliably contain the nodule that we want to connect to. I guess because of snapping it's possible that the mouse is actually hovering over the node and we get that gadget instead.

johnhaddon commented:
> Urg, without having tried yet - I guess what I was missing when making the suggestion to handle it all on the connection side was that we will have to have a connection for that to work. It'll be fine if we just move the connection over to a different plug, but making the first connection always needs to be handled in the Nodule, right?

Yep, I don't see any other way. But I don't think it's any worse than what we have already.

> One more thing that I've noticed: The event.destinationGadget doesn't reliably contain the nodule that we want to connect to. I guess because of snapping it's possible that the mouse is actually hovering over the node and we get that gadget instead.

OK. We should treat that as a bug - we need to guarantee that `event.destinationGadget` as seen by `dragEnd()` is identical to the Gadget that `drop()` was delivered to. That's a fundamental part of the drag+drop protocol. Let me know if you'd like me to take a look at that part...

mattigruener commented:
Hey John,

I have to run out, but I tried to quickly put up something that we can discuss tomorrow. None of this is particularly pretty at the moment, but I think it works.

1. I've "fixed" that Event bug by providing a template specialization that transfers the overwritten destinationGadget back onto the original Event. I've just added it so that I could keep working until we have a better solution (or decide that this is actually what we want). This addresses the eMail I've sent earlier today.

2. I changed the code so that it's the Gadget that the drag was initiated from that is responsible for making the connection (it's done in dragEnd() instead of drop() now). That's kind of nice in the sense that now we have a clear distinction between an initial connection being made (handled by the nodule) and one that is changed by dragging the noodle and which potentially needs to clean up the old connection (handled by the ConnectionGadget).

The code is a little messy at the moment. Will clean it up tomorrow and add some stuff that I'm currently ignoring (checking acceptsInput() for example), but I wanted to get your opinion on whether you can see this ending up what we are going with. Am I missing something?

Thanks!

Matti.

mattigruener commented:
Hey John,

here is a late update with the result of adding that new abstraction that we talked about - and some other stuff. In the order that the commits are currently in:

1. I noticed that the `PlugAdders` on `Boxes` behave really oddly when trying to drag from the little plus icon to plugs on nodes like `Group`. After some detective work, I figured out that the problem is that the `PlugAdder` blindly accepts all plugs, including `ArrayPlugs` which causes the nodule snapping system in `StandardNodeGadget` for `Group` to think that we should snap to the `ArrayPlug`'s `Gadget` as it's the closest (I'm assuming it lives on the whole top edge?). I've just told the Box's PlugAdder here that it should ignore those plugs and added a comment regarding direction, too. I find it a little odd that I can connect both Direction::In and Direction::Out Plugs to the same edge on Boxes, but maybe that's just me.

2. I've left the fix for the violation of the drag&drop protocol in there for now as a reminder. While we can work around the problem with the new abstraction, maybe we should still think about fixing the problem?

3. The following commits are for adding the abstraction and providing implementations for `createConnection()`. I think the implementation and the history should be somewhat clean. I have two questions: 
   - The UndoScope - who would you consider being responsible for setting that up? The caller that uses createConnection() or the implementation of createConnection()? I think intuitively I would think it's the caller that manages the Undo (like I've done in the current implementation), but we also set up the UndoScope already in at least one implementation of the PlugAdders' addPlug() which I have abused here to make the implementation of createConnection() simple. Maybe it doesn't matter, but I was curious to hear your thoughts. I'm happy to add UndoScopes to the individual implementations of createConnection() instead.
   - How to handle `CompoundNodule`? It needs to provide an implementation of createConnection() and updateDragEndPoint() and currently I'm doing that by adding an empty implementation in `Nodule` instead of adding it in `CompoundNodule`. How do you handle that usually? Assuming we add some other *Nodule at some point that doesn't need to provide this functionality, adding empty implementations of the base might make sense? I believe that's what was happening for updateDragEndPoint() already.

4. Adding the shift-click feature was pretty natural after everything in (3) was done. I think this is A LOT better than what we had before. What do you think?

That looks like lots to discuss, but I think we're pretty close? :)

Thanks,

Matti.

johnhaddon commented on include/GafferUI/PlugAdder.h:
```diff
@@ -37,24 +37,32 @@
 		virtual Imath::Box3f bound() const;
 
-		void updateDragEndPoint( const Imath::V3f position, const Imath::V3f &tangent );
+		/// May be called by the recipient of a drag to set a more appropriate position
```
There's no need to repeat the documentation from the base class - it just makes it extremely likely that the two will get out of sync. I think there are a few spots where this applies.

johnhaddon commented on include/GafferUI/StandardConnectionGadget.h:
```diff
@@ -117,6 +117,12 @@ class StandardConnectionGadget : public ConnectionGadget
 		Imath::V3f m_dotPreviewLocation;
 
+		bool m_addingConnection;
+		Imath::V3f m_srcPosOrig;
```
Do we need all four of these V3fs? It seems like two might be sufficient since there's only a new endpoint for the dragged connection - the source point is always shared. We make quite a lot of these gadgets, and they already have a fairly big footprint, so it would be nice to minimise it if possible...

johnhaddon commented on src/GafferUI/StandardConnectionGadget.cpp:
```diff
@@ -195,11 +195,14 @@ void StandardConnectionGadget::createConnection( Gaffer::Plug *destinationPlug )
+			// possibly including the input plug we want to connect to. See issue #302.
+			if( !m_addingConnection )
+			{
+				dstNodule()->plug()->setInput( 0 );
```
nullptr

mattigruener commented:
Hey John,

thanks for the review - will jump onto addressing your feedback this morning.

> > I find it a little odd that I can connect both Direction::In and Direction::Out Plugs to the same edge on Boxes, but maybe that's just me.
> 
> I find it a bit odd too, but I wanted to give users the freedom to design nodes similar to the ones we already have - with an input filter on the right for instance. Or are you suggesting that once a side has a plug, it should reject all connections for plugs with a different direction? Something to think about, but not something we need to deal with here I think.

Yeah, that's what I was thinking. I don't think a node should allow connections of different directions on the same side. That feels very confusing to me. Let's address that later :)

mattigruener commented:
Hey John,

I explored the canCreateConnection() idea as you suggested. Could you take a look and tell me what you think? Seems pretty good to me. 

I used it to clean up some code in StandardNodeGadget. I think that shouldn't affect behaviour too much, except that now snapping isn't supported anymore when dragging from a plug on a node to a PlugAdder on the same node. I noticed that before I changed the code snapping worked and I assumed that we generally want to discourage those connections? People can still drag to the Plug/PlugAdder directly because that's handled elsewhere, but at least we're not helping by providing the snapping? Snapping to Plugs on the same Node was disabled already before I made the changes.

For the PlugAdder I've currently implemented the canCreateConnection() method via acceptsPlug(). I'm happy to remove the acceptsPlug() method entirely, but I wasn't sure if we might want to keep it? It's a little confusing to keep it, I guess, but the bit of code I've changed in StandardNodeGadget makes use of it because it doesn't know the nodule for the given Plug. Maybe that example shows that it's valuable to have a method that takes a Plug as input instead?

I addressed all of the other feedback as well, I think. I squashed in some of the changes directly, for others there's a commit that I will cleanup later. Hopefully all the clang fixes just work now - I just brought over your changes from the branch you posted. Thanks :)

Oh and there's a commit that prevents a call to closestDragDestinationProxy() when starting to drag the node itself. That call probably wasn't that expensive anyway, but it's probably good to avoid the overhead?

Toodle-oo,

Matti.

johnhaddon commented on include/GafferUI/ConnectionCreator.h:
```diff
@@ -63,10 +58,14 @@ class ConnectionCreator : public Gadget
 
+		/// May be called by the recipient of a drag to figure out if this
+		/// ConnectionCreator can set up the connection.
+		virtual bool canCreateConnection( const GafferUI::Nodule *endpoint ) = 0;
```
Where possible I like to order methods in the order of anticipated use. This makes them a bit easier to document and understand. Since the first thing a drag recipient will do is call `canCreateConnection()`, it would be a tiny bit clearer to put this before `updateDragEndPoint()`.

johnhaddon commented on src/GafferUI/StandardNodule.cpp:
```diff
@@ -118,9 +118,30 @@ void StandardNodule::updateDragEndPoint( const Imath::V3f position, const Imath:
+	}
+}
+
+void StandardNodule::createConnection( GafferUI::Nodule *endpoint )
```
Now we have `canCreateConnection()`, there's a lot of unnecessary defensive code in this function that will hide bugs by silently doing nothing. It's the responsibility of the caller to verify `canCreateConnection()` before calling `createConnection()`, so we can simplify this down to just the following :

```
if( localPlug->direction() == Gaffer::Plug::Direction::In )
{
    localPlug->setInput( destinationPlug );
}
else
{
 	destinationPlug->setInput( localPlug );
 }
```

Any rogue callers will now be alerted to their error by an exception...

johnhaddon commented on include/GafferUI/Nodule.h:
```diff
@@ -66,8 +66,6 @@ class Nodule : public ConnectionCreator
 		const Gaffer::Plug *plug() const;
 
 		void updateDragEndPoint( const Imath::V3f position, const Imath::V3f &tangent ) override;
-		bool canCreateConnection( const GafferUI::Nodule *endpoint ) override;
```
Let's do the same with `updateDragEndPoint()` as well, since it's part of the same API.

johnhaddon commented on src/GafferUI/StandardNodeGadget.cpp:
```diff
@@ -573,6 +573,14 @@ void StandardNodeGadget::leave( Gadget *gadget )
 
 bool StandardNodeGadget::dragEnter( GadgetPtr gadget, const DragDropEvent &event )
 {
+	// if the sourceGadget is not capable of making a connection anyway, there's
```
I'm going to suggest that second-guessing the implementation of `closestDragDestinationProxy()` like this is a bad idea. That method has one responsibility - to tell us if we own a Gadget that could be a proxy for a given DragDropEvent. By effectively copying part of those internals here, we're making ourselves vulnerable to changes in the internal implementation, and breaking encapsulation by assuming we know how that function will operate.

I'm pretty sure this is also masking a bug in `closestDragDestinationProxy()`, whereby it calls `creator->canCreateConnection()` on a potentially null creator.

This isn't performance-critical code, and we're now actually doing the test _twice_ in many cases now, once inside and once outside.

johnhaddon commented on src/GafferUI/StandardNodeGadget.cpp:
```diff
@@ -645,6 +645,13 @@ Gadget *StandardNodeGadget::closestDragDestinationProxy( const DragDropEvent &ev
 
+	// gadgets on the same node are invalid destinations and don't support snapping
+	ConnectionCreator *creator = IECore::runTimeCast<ConnectionCreator>( event.sourceGadget.get() );
+	if( creator && isAncestorOf( creator ) )
```
Should this be `if( !creator || isAncestorOf( creator ) )`? Below we are definitely assuming that `creator` is non-null...

mattigruener replied:
Testing github. Please Ignore.

mattigruener commented:
Hey John, - thanks :)

> Looking through what we've got now, I'm wondering if my suggestion of passing Nodule *endpoint was wrong anyway, and we should just be passing Plug *endpoint - do you see any reason why not?

I'm using it in `PlugAdder::canCreateConnection()` to make sure we don't connect to `CompoundNodules`, but I'll have a look if I can get a hold of the plug's nodule in some other way.

Cool, I look forward to talking to you later then.

mattigruener commented:
> It seems like the check for the nodules is being performed in StandardNodule::canCreateConnection(), but the check for the PlugAdders is being done manually in StandardNodeGadget::closestDragDestinationProxy() - is that correct? It feels like maybe we should be consistent there and do the check in PlugAdderDerivedClass::canCreateConnection()? 

Not sure if I understand you correctly, but the check for snapping to either Plugs or PlugAdders is currently done by this code in `StandardNodeGadget::closestDragDestinationProxy()`:

```
	if( creator && isAncestorOf( creator ) )
	{
		return nullptr;
	}
```

I guess that you're suggesting removing that bit and rely solely on `canCreateConnection()`, right? If we want to use `PlugAdder::canCreateConnection()` for that we'd have to see if we can get a hold of the nodule for the plug in `event.data` and then pass that to `canCreateConnection()` called on the `PlugAdder` that lives on the node, right? I find that bit of code not very intuitive to be honest, because we're checking if we can connect TO a `ConnectionCreator` and not if the dragged `ConnectionCreator` can make a connection to the destination of the drag.

> Would that also fix the direct-drag problem you mentioned?

So, we'd still allow a call to StandardNodeGadget::closestDragDestinationProxy() when we start dragging the node itself?

I'll hold off on this for now until we get a chance to have a quick chat tomorrow. Thanks :)

johnhaddon commented:
> Not sure if I understand you correctly, but the check for snapping to either Plugs or PlugAdders is currently done by this code in `StandardNodeGadget::closestDragDestinationProxy()`:

There's also this in `StandardNodule::canCreateConnection()` though, where we're already rejecting "same node" connections :

```
if( localPlug->node() == destinationPlug->node() || localPlug->direction() == destinationPlug->direction() )
{
    return false;
}
```

> I guess that you're suggesting removing that bit and rely solely on `canCreateConnection()`, right?

Yep. But also I'm wondering if we can entirely get rid of the different treatment of PlugAdder and Nodule, since the point of ConnectionCreator was to stop us from having to know about the derived classes :

```
		/// \todo It might be nice if we had a way to connect TO a ConnectionCreator
		if( const Nodule *nodule = IECore::runTimeCast<const Nodule>( it->get() ) )
		{
			if( !creator->canCreateConnection( nodule ) )
			{
				continue;
			}
		}
		else if( const PlugAdder *plugAdder = IECore::runTimeCast<const PlugAdder>( it->get() ) )
		{
			Gaffer::Plug *plug = IECore::runTimeCast<Gaffer::Plug>( event.data.get() );
			if( !plug || !plugAdder->acceptsPlug( plug ) )
			{
				continue;
			}
		}
```

> If we want to use `PlugAdder::canCreateConnection()` for that we'd have to see if we can get a hold of the nodule for the plug in `event.data` and then pass that to `canCreateConnection()` called on the `PlugAdder` that lives on the node, right? I find that bit of code not very intuitive to be honest, because we're checking if we can connect TO a `ConnectionCreator` and not if the dragged `ConnectionCreator` can make a connection to the destination of the drag.

There's nothing in the ConnectionCreator API to suggest that we shouldn't use it to make such a "reverse" connection though. And I think we might even be able to remove the hacky forwarding of events to `m_dragDestinationProxy`, and just call `m_dragDestinationProxy->createConnection( event.data )` in `StandardNodeGadget::drop()`. That would really be the icing on the cake - we'd then have removed every last hack from the connection management code.

I'm now thinking we should definitely pass `Plug *` and not `Nodule *` to ConnectionCreator methods. We'll deal with the CompoundNodule issue some other way - we could query the plug metadata to see if it has a CompoundNodule for instance.

> > Would that also fix the direct-drag problem you mentioned?

> So, we'd still allow a call to StandardNodeGadget::closestDragDestinationProxy() when we start dragging the node itself?

Yes. By "direct-drag" I was referring to another problem I thought you'd mentioned, about dragging directly to a PlugAdder (rather than via the StandardNodeGadget snapping). Perhaps I've got my wires crossed.

> I'll hold off on this for now until we get a chance to have a quick chat tomorrow. 

Cool. Chat to you later...

mattigruener commented:
Hey John,

spent some time today finishing the snapping refactoring and cleaning up the history - I _think_ it should be pretty clean now. I'll have a look at those conflicts that github is showing tomorrow. Time to go home now. I hope my tired brain didn't mess up anything, but it seems to still work as I would expect.

Thanks!

Matti.

mattigruener commented:
Oh, I extended the bindings for `ConnectionCreator` and adjusted existing `DerivedPlugAdders` too - I guess that's gonna need a review.

johnhaddon commented:
Thanks Matti,
I grabbed a copy of this for testing, and in doing so fixed up the conflicts. My branch is here :

https://github.com/johnhaddon/gaffer/commits/mattigruener-shiftClickConnection

I've also added some further commits for discussion. Some of them are obviously fine I think, but others might be more contentious. When you're back online let's discuss how we get this over the finish line...
Cheers...
John

mattigruener commented:
Hey John,

I finally wanted to get back onto this one to get it over the finish line, but I ran into an issue when trying to compile the branch in your repo. Is it possible that you forgot to commit GafferUIBindings/ConnectionCreatorBinding.inl? It's included in the header, but -unless I messed something up- I don't think it exists. 

Thanks

Matti.

mattigruener commented:
Hey John,

I gotta run but wanted to have something up for you to look at. It's not pretty and needs some cleanup in some areas. The biggest issue is a weird dragMove behaviour:

If you create two Group nodes, drag from the first's out plug onto the label that appears for that plug, it'll think it should connect to the second node. I'm not sure if I caused that or just uncovered a deeper bug. Some gdb investigation shows that the dispatch for that misguided signal is coming from ViewportGadget.cpp:621. Maybe I'm failing to initialize the data somewhere or it never has been done properly and I've just uncovered it with my additional rendering from the nodes... I'm just guessing here as I don't have time to investigate further. I'll continue tomorrow if you think that the general direction is alright.

I've also taken care of the tangent bits that were missing. It's not pretty and I had to change the PlugAdder API slightly. Do you see a better way?

Thanks! Let's talk tomorrow as planned :)

Cheerio!

johnhaddon commented:
> If you create two Group nodes, drag from the first's out plug onto the label that appears for that plug, it'll think it should connect to the second node. I'm not sure if I caused that or just uncovered a deeper bug.

I'm pretty sure this is a new one. If you look at the implementation for Gadget, you'll see that it's designed such that the public `Gadget::render()` method does some bookkeeping and then delegates to the protected virtual `doRender()` to do the actual drawing. [Part of that bookkeeping](https://github.com/GafferHQ/gaffer/blob/master/src/GafferUI/Gadget.cpp#L277) is to support hit-testing of the gadget under the mouse. The `doRenderOverlays()` stuff that you've added isn't doing any of that bookkeeping, so the hit-testing is getting confused, and selecting the last thing that did do the bookkeeping correctly - the `render()` method for the second node gadget.

There are some additional issues with the `doRenderOverlay()` approach as it is so far :

- It's totally specific to node gadgets, but other gadgets might like to have overlays
- It doesn't address the other problem we have, which is always rendering some things _under_ everything else (which is currently dealt with by the special-case code for connections in `GraphGadget::doRender().

One option would be to move everything to Gadget/ViewportGadget and add equivalent `renderUnderlay()` methods but I wasn't sure if that was sufficient as a general mechanism, which is why I punted and suggested using an extra gadget as the overlay. Maybe it's worth figuring this out though, as it'd definitely be useful for other stuff.

We can chat about this later, I just wanted to get some thoughts down while they were fresh...

johnhaddon commented:
> other gadgets might like to have overlays

Here's one example :

<img width="360" alt="nodulelayering" src="https://user-images.githubusercontent.com/1133871/31827507-dcd46040-b5af-11e7-8c70-ec703c8be045.png">

It's a bit annoying that the larger "hovered" nodules draw on top of the nodule above, and below the nodule below. If we instead drew that as an overlay it'd be on top of both...

mattigruener commented:
Papertrail to help navigating our rabbit hole:

I've added https://github.com/GafferHQ/gaffer/pull/2304 to address some of the concerns about rendering under and over Gadgets in a more general manner. This PR will build on that one.

mattigruener commented:
I have rebased onto the PR that introduces layered Gadget rendering. Those commits should disappear here once that PR is merged. I believe the history is a little messed up at the moment, but I'll hold off cleaning it up until that other PR is in master.

mattigruener commented:
Rebased onto the new branch that handles Gadget rendering. Will continue here once that branch is merged. I believe the only outstanding issue of the ones you mentioned in your branch, John, is the additional method that needs implementing when creating a PlugAdder in Python?

So close! ;)

mattigruener commented:
Grepping tells me that there's no python based PlugAdder in Gaffer and we only have one or two internal projects where we use those. I can fix up those pretty quickly once we merged this PR. Can you think of anything else that needs to be done here?

johnhaddon commented:
Hey Matti, could you rebase this now we have everything merged?
Cheers...
John

mattigruener commented:
I thought I had done so already? I just pushed my local branch again that didn't show any updates when rebasing again... This should have everything needed now? :)

johnhaddon commented:
Hey Matti,
Sorry for the delay on this. I'm keeping it parked for a bit while I explore the various issues that the layered rendering have thrown up. I'm not expecting any more from you though, so please just consider the ball to be in my court, and point any disgruntled would-be-shift-draggers in my direction...
Cheers...
John

mattigruener commented:
Heyo :)

Yeah, that's what I thought - no worries. Please feel free to throw the ball back into my court on those issues if you feel like I could/should clean up some of that performance stuff. I still haven't had a chance to look through that original PR again, but I got it on my list of things to do.

Thanks for the update :)

mattigruener commented:
I think you mentioned that we could probably merge this? Then I could cross it off my list, but no worries if we want to wait a bit longer :)

mattigruener commented on include/GafferUI/CompoundNodule.h:
```diff
@@ -38,6 +38,7 @@
 #define GAFFERUI_COMPOUNDNODULE_H
 
+#include "Gaffer/Plug.h"
 #include "GafferUI/Nodule.h"
```
foo

johnhaddon replied:
Since you're here, I should mention that I'm finally in the process of getting this branch merged. I've dealt with all the conflicts, and I'm now just agonising over some details of the drag proxy and labelling stuff. Once I'm done I'll hand back to you for a final review/test.

johnhaddon commented:
Closing in favour of #2480.
